### PR TITLE
WarFile.java:147-148: Implement validation of JSON inside... #16

### PR DESCRIPTION
**Pull request by nhekfqn**

### CODE REVIEW

**Comment by dmarkov** [comment]
Let me find a reviewer for this pull request, thanks for submitting it

**Comment by dmarkov** [comment]
@krzyk please review

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 1]
```diff
@@ -100,4 +103,32 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
             );
         }
     }
+
+    /**
+     * Verifies that checkEbextensionsValidity runs fine
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: - *WarFile can use a .ebextensions with valid json*

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 2]
```diff
@@ -100,4 +103,32 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+     * when .ebextensions entry has a valid json text.
+     * @throws Exception Thrown in case of error.
+     */
+    @Test
+    public void checkValidateJson() throws Exception {
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: *usesEbextensionsWithValidJson*

File: src/main/java/com/jcabi/beanstalk/maven/plugin/WarFile.java [hunk 3]
```diff
@@ -144,13 +146,14 @@ private boolean validYaml(final String text) {
...
-     * @todo #2:30min Implement validation of JSON inside the method
-     *  AbstractBeanstalkMojo.validJson(). Remember to unit test your solution.
      */
     private boolean validJson(final String text) {
-        throw new NotImplementedException(
-            "com.jcabi.beanstalk.maven.plugin.AbstractBeanstalkMojo.validJson"
-        );
+        boolean result = true;
+        try {
+            new JSONObject(text);
```
Comment (krzyk) [inline src/main/java/com/jcabi/beanstalk/maven/plugin/WarFile.java]: text could be also a `JSONArray`

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 4]
```diff
@@ -100,4 +103,32 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+        Mockito.when(enumeration.nextElement()).thenReturn(entry)
+            .thenReturn(null);
+        Mockito.when(entry.getName()).thenReturn(".ebextensions/");
+        Mockito.when(zip.getInputStream(entry)).thenReturn(
+            new ByteArrayInputStream(
+                "{\"blah\":\"blah\"}".getBytes(StandardCharsets.UTF_8)
+            )
+        );
+        final WarFile war = new WarFile(zip);
+        war.checkEbextensionsValidity();
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: this test doesn't assert or verify anything, either add that or add information in the test description that the only verification is lack of exception

**Comment by krzyk** [comment]
@nhekfqn few comments above, and you should add a test for invalid json also

**Comment by nhekfqn** [comment]
@krzyk please review.

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 5]
```diff
@@ -100,4 +105,95 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+    }
+
+    /**
+     * Prepares ZipFile mock containing .ebextensions entry with text
+     * provided.
+     * @param text Text .ebextensions entry contains.
+     * @return ZipFile mock.
+     * @throws IOException Thrown in case of error.
+     */
+    private ZipFile zipWithEbextensionsWithText(final String text)
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: just `zipWithEbextensions(final String text)` is enough

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 6]
```diff
@@ -100,4 +105,95 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+    public void throwsExceptionWhenUsesEbextensionsWithInvalidJson()
+        throws Exception {
+        final ZipFile zip = this.zipWithEbextensionsWithText(
+            "{\"blah\":\"blah\"]"
+        );
+        final WarFile war = new WarFile(zip);
+        try {
+            war.checkEbextensionsValidity();
+            Assert.fail();
+        } catch (final MojoFailureException exception) {
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: you could just add `@Test(expected = MojoFailureException.class)` so you won't have to catch it here. The test will fail if given exception is not thrown.

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 7]
```diff
@@ -100,4 +105,95 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+     * contains invalid json.
+     * @todo #16 enable test when WarFile.validYaml implemented
+     * @throws Exception Thrown in case of error.
+     */
+    @Test
+    @Ignore
+    @SuppressWarnings("StringLiteralsConcatenationCheck")
+    public void throwsExceptionWhenUsesEbextensionsWithInvalidJson()
+        throws Exception {
+        final ZipFile zip = this.zipWithEbextensionsWithText(
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: inline `zip`

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 8]
```diff
@@ -100,4 +105,95 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
             );
         }
     }
+
+    /**
+     * Verifies that WarFile can use a .ebextensions with valid json object
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: Please drop the *Verifies that* part in all test descriptions

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 9]
```diff
@@ -100,4 +105,95 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+     * string. Test passes when WarFile.checkEbextensionsValidity method
+     * doesn't throw exception.
+     * @throws Exception Thrown in case of error.
+     */
+    @Test
+    public void usesEbextensionsWithValidJsonObject() throws Exception {
+        final ZipFile zip = this.zipWithEbextensionsWithText(
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: in all tests you could just inline `zip` as well as `war`.

**Comment by krzyk** [comment]
@nhekfqn we are almost there :), see my comments above

**Comment by nhekfqn** [comment]
@krzyk 
> you could just add @Test(expected = MojoFailureException.class)
I know, but I couldn't check exception message in that case. MojoFailureException can be thrown from three places in WarFile.checkEbextensionsValidity so its message does matter.

Other defects fixed, please review.

**Comment by krzyk** [comment]
@nhekfqn You can use [ExpectedException](http://junit.org/javadoc/latest/org/junit/rules/ExpectedException.htm) for checking message

**Comment by nhekfqn** [comment]
@krzyk please review.

File: src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java [hunk 10]
```diff
@@ -100,4 +114,89 @@ public void checkEbextensionsValidityThrowsExceptionNoConfigFiles()
...
+        new WarFile(
+            this.zipWithEbextensions(
+                "[{\"blah\":\"blah\"},{\"blah\":\"blah\"}]"
+            )
+        ).checkEbextensionsValidity();
+    }
+
+    /**
+     * WarFile throws exception when .ebextensions contains invalid json.
+     * @todo #16 enable test when WarFile.validYaml implemented
```
Comment (krzyk) [inline src/test/java/com/jcabi/beanstalk/maven/plugin/WarFileTest.java]: `@todo #16:15min Enable this test when WarFile.validYaml is implemented.`

**Comment by krzyk** [comment]
@nhekfqn a comment about todo formatting

**Comment by nhekfqn** [comment]
@krzyk Please review.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by dmarkov** [comment]
@krzyk done, I added **17 mins** in payment `51683416`... extra minutes for review comments (c=2)... added +17 to your rating, now it is equal to [+4210](http://www.netbout.com/b/35035?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Pull request merged**